### ISSUE TITLE
Kopia doesn't automatically run overdue snapshots when waking from sleep

### ISSUE THREAD

**Issue by elydpg** [kopia/kopia#3512]
I noticed that #3319 is supposed to fix overdue snapshots not automatically running. And it does. However, overdue snapshots do not run when the computer was put to sleep (with Kopia running of course) before the designated snapshot time and woken from sleep after the snapshot was supposed to have been run. Instead, Kopia needs to be restarted in order for the missed snapshot to be automatically run.

For further info, I am running Kopia 0.15.0 on Kubuntu 23.10.

**Comment by elydpg** [comment]
Kopia upon waking from system sleep, before restarting Kopia

![Screenshot_20231217_012103](https://github.com/kopia/kopia/assets/8622264/2f2f3f44-693b-497e-941d-bb5842620ad7)

After restarting Kopia

![Screenshot_20231217_012353](https://github.com/kopia/kopia/assets/8622264/24e93158-d25f-430e-8b31-a8f91ec29ba2)

**Comment by lupusA** [comment]
Hi, 

can you post your policy?

Cheers

**Comment by elydpg** [comment]
No problem! My policy is specified by this JSON representation:

```
{
    "retention": {},
    "files": {
        "ignore": [
            "[censored]"
        ]
    },
    "errorHandling": {},
    "scheduling": {
        "timeOfDay": [
            {
                "hour": 4,
                "min": 0
            }
        ]
    },
    "compression": {},
    "actions": {},
    "logging": {
        "directories": {},
        "entries": {}
    },
    "upload": {}
}
```

The global policy that it inherits from is defined by this:

```
{
    "retention": {
        "keepLatest": 10,
        "keepHourly": 48,
        "keepDaily": 7,
        "keepWeekly": 4,
        "keepMonthly": 24,
        "keepAnnual": 3,
        "ignoreIdenticalSnapshots": false
    },
    "files": {
        "ignoreDotFiles": [
            ".kopiaignore"
        ]
    },
    "errorHandling": {
        "ignoreFileErrors": false,
        "ignoreDirectoryErrors": false,
        "ignoreUnknownTypes": true
    },
    "scheduling": {
        "runMissed": true
    },
    "compression": {
        "compressorName": "none"
    },
    "actions": {},
    "logging": {
        "directories": {
            "snapshotted": 5,
            "ignored": 5
        },
        "entries": {
            "snapshotted": 0,
            "ignored": 5,
            "cacheHit": 0,
            "cacheMiss": 0
        }
    },
    "upload": {
        "maxParallelSnapshots": 1,
        "parallelUploadAboveSize": 2147483648
    }
}
```

**Comment by tessus** [comment]
The same happens on macOS. My workaround is to click the refresh button: 
<img width="49" alt="image" src="https://github.com/kopia/kopia/assets/223439/c11d43e0-5947-43ac-9912-c1ac74d9cbf7">

**Comment by elydpg** [comment]
Still happens on 0.16.0

**Comment by blm768** [comment]
I've confirmed similar behavior on Windows in 0.17.0. I only see the behavior on a repository that's stored on an SMB share; my other machine that backs up to an S3-compatible object store doesn't exhibit the issue. Kopia does seem to try to run the snapshots, but they fail (perhaps because the SMB share isn't ready yet). This also seems to block later hourly snapshots from running until I refresh the UI.

**Comment by DarkArc** [comment]
I'm seeing this as well on my laptop running Fedora 40.

**Comment by manuel-koch** [comment]
I have the same issue on MacOS Apple M3 Pro, running kopia 0.19, on a remote SFTP repo.
Having setup a schedule of "every 15min" I see that this seems to execute backup attempts at the correct interval, but only until I close my laptop. Once I open the laptop again, the schedule is broken and I need to trigger a backup on all snapshots manually to get the schedule running again.

**Comment by gsupp** [comment]
I can confirm similar behavior with KopiaUI v0.20.1 on Linux Mint 21.3, backing up to a Kopia Repository Server. I have a snapshot scheduled to run every 15 minutes, which it faithfully does until the laptop is suspended long enough to miss the next scheduled time. On resume, KopiaUI shows the Next Snapshot as overdue but doesn't seem to execute the snapshot until hitting the refresh button. 

![Image](https://github.com/user-attachments/assets/b9eb8e51-6c49-4b5d-b26b-d85608df5dd5)

**Comment by jkowalski** [comment]
Thanks for the report. I'll try to reproduce it and make a fix.

**Comment by jkowalski** [comment]
I tried to reproduce this on my macOS in the latest KopiaUI and I get this behavior:

1. set snapshot interval to 2 minutes (`kopia policy set ~/Documents --snapshot-interval 120s`)
2. at 7:40:00 the snapshot was created
3. at 7:40:30 I've put my laptop to sleep
4. woke up around 7:50:05
5. the snapshot was marked as overdue and stayed like this for ~2 minutes
6. the snapshot was created at the next interval on 7:52:00

Is this what you're seeing?

**Comment by DarkArc** [comment]
I suspect it is. However, for daily backups scheduled at hours when the device may not be used (e.g., 1am -- since 1am is likely consistently missed by the laptop sleeping) the backup may not run until the next reboot (when KopiaUI takes a fresh look at the situation and schedules the overdue run).

Perhaps having a 5-15 minute "pulse" to check for overdue snapshots and catch up would resolve the issue?

**Comment by gsupp** [comment]
@jkowalski, you're right, that is the behavior I am seeing as well. Snapshots that are missed while the computer is suspended are not caught up on once woken up, but they do run at the next scheduled interval. Steps:

1. set 10 minute snapshot interval
2. snapshot ran at 12:50:00 PM
3. suspended computer at 12:55 PM
4. woke computer at 1:05 PM
5. snapshot ran at 1:11:48 PM

**Comment by jkowalski** [comment]
what I think is happening is:

1. kopia determines that next event is 10 minutes from now
2. kopia starts sleeping for 10 minutes
3. let's say laptop goes to sleep after 1 minute
4. long time passes and when the laptop wakes up, the previous sleep still has 9 minutes to go...
5. kopia continues to sleep for 9 minutes
6. at this point we re-evaluate future snapshot times and that's why the next snapshot will generally run

that likely explains why your snapshot ran at 1:11:48PM and not at 1:10:00PM

I think the fix would be to limit the sleep duration to something like 1 minute, so after waking up from sleep the timer will complete in <1min and will re-evaluate upcoming snapshots promptly.

**Comment by tessus** [comment]
My scenario is a bit different, which is why I wanted to mention it:

I have the following in my policy `Times Of Day: 21:00`. What happens is that when my MacBook is suspended before 9pm and also woken up before 9pm (e.g. 8:37pm), the backup is not run at 9pm and only shows `overdue`. When I click the "refresh" button, the backup starts.

**Comment by tessus** [comment]
Unfortunately something seems to be very wrong with this fix. I installed 0.21.0 yesterday and my backup is scheduled to run at 21:00. The last day it ran at 21:00 so I thought it would start again at 21:00. But as soon as I started 0.21.0 a backup was run and when it was done it started a new backup. Again and again. I had to manually stop this backup loop.
And today around 17:00 it started again. I am currently at the 3rd loop:

<img width="729" height="358" alt="Image" src="https://github.com/user-attachments/assets/2459a770-00dd-4627-8eed-084c70797d56" />

I have now stopped the backup and disabled the automatic backup completely.

**Comment by tessus** [comment]
<img width="743" height="487" alt="Image" src="https://github.com/user-attachments/assets/38799e4c-deaf-4fe3-8683-15a6c15a84f5" />

Yes, please let me know, if I should open a new issue. @jkowalski @julio-lopez

**Comment by tessus** [comment]
just found a new issue #4729. I'll put my 2 comments there.
